### PR TITLE
Transductive Learning

### PR DESCRIPTION
**Pull request by jwmueller**
Adds basic version of pseudolabeling specifically designed for test data. The new argument added is `unlabeled_test_data` as the transductive setting is sufficiently distinct from the semi-supervised setting, especially as we add shift-detection and other test-data-specific functionality, it will be crucial to have a dedicated argument that makes it clear the resulting models are specifically designed for the test data (rather than any new data).  This also avoids merge conflict with the TabTransformer PR's use of `unlabeled_data` argument.

Example code:

```
from autogluon import TabularPrediction as task

label_column = 'class'

train_data = task.Dataset(file_path='https://autogluon.s3.amazonaws.com/datasets/Inc/train.csv')
val_data = train_data[1000:1500]
train_data = train_data.head(500)
test_data = task.Dataset(file_path='https://autogluon.s3.amazonaws.com/datasets/Inc/test.csv')
test_data = test_data.head(5000)
y_test = test_data[label_column]
test_data_nolab = test_data.drop(labels=[label_column],axis=1)

hyperparameters = {'GBM': {},'CAT':{}}  # Just here to ensure quicker run

# Transductive learning w validation data:
predictor = task.fit(train_data=train_data, tuning_data=val_data, unlabeled_test_data=test_data_nolab, label=label_column, hyperparameters=hyperparameters)

# Transductive learning:
predictor = task.fit(train_data=train_data, unlabeled_test_data=test_data_nolab, label=label_column, hyperparameters=hyperparameters)

# Transductive learning w stacking:
predictor = task.fit(train_data=train_data, unlabeled_test_data=test_data_nolab, label=label_column, auto_stack=True, hyperparameters=hyperparameters)


y_pred = predictor.predict(test_data_nolab)
perf = predictor.evaluate_predictions(y_true=y_test, y_pred=y_pred)
predictor.leaderboard(test_data)


## Transductive regression:
label_column = 'age'
y_test = test_data[label_column]
test_data_nolab = test_data.drop(labels=[label_column],axis=1) # delete labels from test data since we wouldn't have them in practice

predictor = task.fit(train_data=train_data, unlabeled_test_data=test_data_nolab, label=label_column, hyperparameters=hyperparameters)
predictor.leaderboard(test_data)


## Transductive multiclass classification:
label_column = 'native-country'
y_test = test_data[label_column]
test_data_nolab = test_data.drop(labels=[label_column],axis=1) # delete labels from test data since we wouldn't have them in practice

predictor = task.fit(train_data=train_data, unlabeled_test_data=test_data_nolab, label=label_column, hyperparameters=hyperparameters)
predictor.leaderboard(test_data)
```

### CODE REVIEW

**Comment by szha** [comment]
Job PR-668-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-668/1/index.html

**Comment by szha** [comment]
Job PR-668-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-668/2/index.html

File: autogluon/task/tabular_prediction/predictor.py [hunk 1]
```diff
@@ -91,6 +91,7 @@ def __init__(self, learner):
         self.class_labels = self._learner.class_labels
         self.class_labels_internal = self._learner.label_cleaner.ordered_class_labels_transformed
         self.class_labels_internal_map = self._learner.label_cleaner.inv_map
+        self.fit_transductively = self._learner.fit_transductively
```
Comment (daveharmon) [inline autogluon/task/tabular_prediction/predictor.py:94]: Add documentation for this arg with the other arguments

**Review by daveharmon — COMMENTED** [review]
There should probably be new tests with these features enabled

File: autogluon/utils/tabular/ml/learner/default_learner.py [hunk 2]
```diff
@@ -105,7 +117,120 @@ def _fit(self, X: DataFrame, X_val: DataFrame = None, scheduler_options=None, hy
...
+            save_data=save_data,
+            save_bagged_folds=save_bagged_folds,
+            random_seed=self.random_seed,
+            verbosity=verbosity
+        )
+        self.inductive_trainer_path = self.trainer_path
+        self.trainer_path = trainer.path
+        trainer.train(X, y, X_val, y_val, hyperparameter_tune=hyperparameter_tune, feature_prune=feature_prune, holdout_frac=holdout_frac,
+                      hyperparameters=hyperparameters, ag_args_fit=ag_args_fit, excluded_model_types=excluded_model_types, name_suffix=TRANSDUCTIVE_SUFFIX)
+        self.save_trainer(trainer=trainer)
```
Comment (Innixma) [inline autogluon/utils/tabular/ml/learner/default_learner.py:226]: Making predictions with these new models on the test data can cause severe score drop due to the predictions being generated from models which used the pseudo-labels of the test data during training. This is especially visible in log_loss eval_metric, where RF can be extremely overconfident about its training data predictions, leading to a very bad score when a pseudo-label happened to be different than ground truth.

Tested on Adult with eval_metric=log_loss, RF went from 0.307 logloss (without trans) to 0.378 (with trans). In addition, val_score didn't change much, making the trans trained model's val_score misleading (Both had ~0.306 val score).

I believe that for this to be usable, predictions for test rows that were used in pseudo-labeled train data must come from the original models and not the trans models, or they must be trained with soft_class.

File: autogluon/utils/tabular/ml/learner/default_learner.py [hunk 3]
```diff
@@ -64,6 +67,15 @@ def _fit(self, X: DataFrame, X_val: DataFrame = None, scheduler_options=None, hy
             logger.log(20, f'Tuning Data Columns: {len(X_val.columns)}')
         time_preprocessing_start = time.time()
         logger.log(20, 'Preprocessing data ...')
+        if X_test_unlabeled is not None:
+            joint_feature_generator = copy.deepcopy(self.feature_generator)
+            # TODO: consider caching original data to file to save memory.
+            X_og = X.copy()
```
Comment (Innixma) [inline autogluon/utils/tabular/ml/learner/default_learner.py:73]: Better to add save/load of pkl's for the data like is done in trainer. Should probably be done regardless of if unlabeled data is present.

File: autogluon/utils/tabular/ml/learner/default_learner.py [hunk 4]
```diff
@@ -105,7 +117,120 @@ def _fit(self, X: DataFrame, X_val: DataFrame = None, scheduler_options=None, hy
         time_end = time.time()
         self.time_fit_training = time_end - time_preprocessing_end
         self.time_fit_total = time_end - time_preprocessing_start
-        logger.log(20, f'AutoGluon training complete, total runtime = {round(self.time_fit_total, 2)}s ...')
+        if X_test_unlabeled is not None:
```
Comment (Innixma) [inline autogluon/utils/tabular/ml/learner/default_learner.py:120]: This should be moved out of _fit. This code can't reasonably be at the end of _fit, but should be an entirely new call to learner. (post-fit call). No need to fit a new feature generator, can simply pass the test data as unlabelled data into the original _fit call and have the feature generator fit with it (identical to the existing functionality of passing in tuning_data without label)

X, X_val, y, y_val can be loaded from cached files similarly to how it is done in trainer.

self.trainer can be re-used, no need to create a new trainer.

File: autogluon/utils/tabular/ml/learner/default_learner.py [hunk 5]
```diff
@@ -105,7 +117,120 @@ def _fit(self, X: DataFrame, X_val: DataFrame = None, scheduler_options=None, hy
...
+            else:
+                remaining_time = None
+            self.transductive_fit(X=X_og, X_val=X_val_og, X_test_unlabeled=X_test_unlabeled, scheduler_options=scheduler_options, hyperparameter_tune=hyperparameter_tune,
+                                  feature_prune=feature_prune, holdout_frac=holdout_frac, num_bagging_folds=num_bagging_folds, num_bagging_sets=num_bagging_sets, stack_ensemble_levels=stack_ensemble_levels,
+                                  hyperparameters=hyperparameters, ag_args_fit=ag_args_fit, excluded_model_types=excluded_model_types, time_limit=remaining_time,
+                                  save_data=save_data, save_bagged_folds=save_bagged_folds, verbosity=verbosity, feature_generator=joint_feature_generator)
+        else:
+            logger.log(20, f'AutoGluon training complete, total runtime = {round(self.time_fit_total, 2)}s')
+
+    def transductive_fit(self, X: DataFrame, X_val: DataFrame, X_test_unlabeled: DataFrame,
```
Comment (Innixma) [inline autogluon/utils/tabular/ml/learner/default_learner.py:133]: This method should return the predict_proba of X_test_unlabeled. This will be the only time it is easily possible to generate the predict_proba for X_test_unlabeled without severe overfitting (unless using softclass): See comment on adult benchmark result for context.

File: autogluon/utils/tabular/ml/learner/default_learner.py [hunk 6]
```diff
@@ -105,7 +117,120 @@ def _fit(self, X: DataFrame, X_val: DataFrame = None, scheduler_options=None, hy
...
+                         scheduler_options, hyperparameter_tune, feature_prune, holdout_frac, num_bagging_folds, num_bagging_sets, stack_ensemble_levels,
+                         hyperparameters, ag_args_fit, excluded_model_types, time_limit, save_data, save_bagged_folds, verbosity,
+                         feature_generator, pseudolabel=True, detect_shift=False):
+        """ After initial fit, transductive_fit will continue training additional models using additional unlabeled test data (X_test_unlabeled).
+            Arguments are the same as Learner.fit().
+            Additional arguments:
+                pseudolabel (bool): whether to utilize pseudolabeling
+                detect_shift (bool): whether to correct for distribution-shift between training/test data distributions.
+
+            TODO: add hyperparameter to only retain confident pseudolabels.
```
Comment (Innixma) [inline autogluon/utils/tabular/ml/learner/default_learner.py:143]: without using softclass or thresholding, I don't think this is usable, as it will frequently harm results.

File: autogluon/task/tabular_prediction/tabular_prediction.py [hunk 7]
```diff
@@ -97,6 +98,13 @@ def fit(train_data,
...
             If `tuning_data = None`, `fit()` will automatically hold out some random validation examples from `train_data`.
+        unlabeled_test_data : str or :class:`autogluon.task.tabular_prediction.TabularDataset` or `pandas.DataFrame`, default = None
+            Table of unlabeled test data which contains the examples you want predictions for (same format as training data, except without label column).
+            Should only be specified if this is the only test data you'll want to produce predictions for after training (transductive learning rather than inductive learning).
+            For example, `unlabeled_test_data` may boost accuracy if using AutoGluon in a prediction competition where the particular test dataset has been provided.
+            Provide large time_limits to utilize transductive training, as it will only begin after a standard `fit()` run that does not utilize the test data.
+            Models that have been trained transductively will have suffix '_TRAN' in their names (these models are biased to perform better on your test data than other new data, unlike the other models).
```
Comment (Innixma) [inline autogluon/task/tabular_prediction/tabular_prediction.py:106]: `_TRANS`

**Review by Innixma — CHANGES_REQUESTED** [review]
Before merging, we will have to ensure this doesn't frequently severely harm results, and is stable regarding time limits.

File: autogluon/utils/tabular/ml/learner/default_learner.py (hunk 4 above)
Comment (Innixma) [reply autogluon/utils/tabular/ml/learner/default_learner.py:120]: Actually, are you sure you need X, X_val, y, y_val at all? Does this really need to live in learner? Why not do this in trainer? The X/X_val won't actually be aligned properly at present since X_val is generally None and the split occurs by random seed in trainer. Altering X will alter the rows used as random seed.

**Comment by szha** [comment]
Job PR-668-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-668/3/index.html

File: autogluon/utils/tabular/ml/learner/default_learner.py [hunk 8]
```diff
@@ -105,7 +117,120 @@ def _fit(self, X: DataFrame, X_val: DataFrame = None, scheduler_options=None, hy
...
+
+        if detect_shift:
+            raise NotImplementedError("detect_shift is not yet implemented.")
+        if (not pseudolabel) and (not detect_shift):
+            raise ValueError("One of the arguments: {pseudolabel, detect_shift} must = True")
+
+        noval_nobag = (X_val is None) and (num_bagging_folds == 0)
+        if pseudolabel:
+            logger.log(20, f"Pseudolabeling the test data ...")
+            test_pseudolabels = self.predict(X=X_test_unlabeled)
```
[the remark this answers is not in the document]
Comment (gradientsky) [reply autogluon/utils/tabular/ml/learner/default_learner.py:165]: +1: it'll result in significant over-fit without threshold or row weighting.

File: autogluon/utils/tabular/ml/learner/default_learner.py (hunk 2 above)
Comment (gradientsky) [reply autogluon/utils/tabular/ml/learner/default_learner.py:226]: Pseudo-labeled rows should be excluded from fold validation set when we calculate scores. Otherwise you will extremely over-confident fold metric estimates, which won't generalize.

**Comment by Innixma** [comment]
Closing this for now, please re-open if updates are made to incorporate with latest AutoGluon
